llvm-project/Dockerfile

```diff
@@ -0,0 +1,133 @@
+FROM alpine:edge AS source
+RUN wget --no-clobber \
+  https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.1/llvm-project-14.0.1.src.tar.xz
+
+# Stage one
+FROM alpine:edge AS stage_one
+
+# Install the stage zero compiler; the prebuilt clang from alpine.
+# _Unwind_Resume issues with llvm-libunwind? Cmake can't find it either?
+# TODO: can we build stage1 with llvm binutils? llvm13 alpine package contains.
+#RUN apk add --no-cache clang lld compiler-rt musl-dev
+RUN apk add --no-cache clang lld musl-dev
+
+COPY --from=source llvm-project-14.0.1.src.tar.xz /
+RUN tar xf llvm-project-14.0.1.src.tar.xz && \
+  mv llvm-project-14.0.1.src llvm-project && \
+  mkdir llvm-project/llvm/build
+
+# libcxxabi depends on zlib-dev!
+# compiler-rt depends on libexecinfo-dev! (unless gwp-asan is disabled)
+# TODO: libxml2?
+# llvm cmake throws lots of warnings without git
+RUN apk add --no-cache cmake ninja python3 zlib-dev git
+
+# This seems like a mistake that clang++ in alpine cannot compile C++ hello
+# world.
+# TODO: this also pulls in gnu binutils...
+RUN apk add --no-cache g++ && \
+  find /usr/bin/ | grep -e gcc -e g++ | grep -v clang++ | xargs rm && \
+  rm /usr/bin/cpp /usr/bin/c99 /usr/bin/c89 && \
+  ln -sf /usr/bin/clang /usr/bin/cc
+
+# Test stage 0 clang and clang++.
+COPY hello.c hello.cpp /
+
+RUN clang hello.c && ./a.out && \
+  clang++ hello.cpp && ./a.out
+
+COPY stage1.cmake llvm-project/.
+RUN cmake \
+  -B llvm-project/llvm/build \
+  -C llvm-project/stage1.cmake \
+  -S llvm-project/llvm \
+  -G Ninja
+
+# While these could be one single command, generally for development it's very
+# painful when one of many specified targets fail as then nothing gets cached
+# by docker. Break these into distinct commands to optimize for development
+# iteration speed.
+RUN ninja -C llvm-project/llvm/build install-builtins
+RUN ninja -C llvm-project/llvm/build install-compiler-rt
+RUN ninja -C llvm-project/llvm/build install-unwind
+RUN ninja -C llvm-project/llvm/build install-cxxabi
+# Necessary for libcxx
+RUN apk add --no-cache linux-headers
+RUN ninja -C llvm-project/llvm/build install-cxx
+RUN ninja -C llvm-project/llvm/build install-clang install-lld
+RUN ninja -C llvm-project/llvm/build install-clang-resource-headers
+
+### END STAGE1
+### START STAGE2
+
+# STAGE2 goal, build WITHOUT apk add
+FROM alpine:edge as stage_two
+COPY --from=stage_one llvm-project/llvm/build/bin /usr/local/bin
```

Comment:
Because stage1's /usr/local/bin still contains binutils, cmake, and other dependencies installed previously.  I guess I could (probably should) apk del everything at the end of a stage?

Reply:
Actually, I think Alpine's apk writes to /usr/bin, not /usr/local/bin. Let me triple check my assertion.

Reply:
Right, I would expect all Alpine packages to be in `/usr/bin`, where as `/usr/local/bin` should only contain files that we installed ourselves. Would be good to make sure though.

Reply:
Stopping the build (by commenting out half the Dockerfile) at stage one, then
```
$ docker run -it clangbuiltlinux/llvm-project:latest ash
/ # ls -l /usr/local/bin/
total 139240
lrwxrwxrwx    1 root     root             8 May 19 18:34 clang -> clang-14
lrwxrwxrwx    1 root     root             5 May 19 18:34 clang++ -> clang
-rwxr-xr-x    1 root     root      85718392 May 19 18:27 clang-14
lrwxrwxrwx    1 root     root             5 May 19 18:34 clang-cl -> clang
lrwxrwxrwx    1 root     root             5 May 19 18:34 clang-cpp -> clang
lrwxrwxrwx    1 root     root             3 May 19 18:34 ld.lld -> lld
lrwxrwxrwx    1 root     root             3 May 19 18:34 ld64.lld -> lld
-rwxr-xr-x    1 root     root      56856976 May 19 18:27 lld
lrwxrwxrwx    1 root     root             3 May 19 18:34 lld-link -> lld
lrwxrwxrwx    1 root     root             3 May 19 18:34 wasm-ld -> lld
```
so yeah, looks like I can just grab `/usr/local/bin`.

Reply:
interesting, this exposed that I need to `ninja install` a few more binaries during stage1. For instance, llvm-ar.
```
/ # ls -l llvm-project/llvm/build/bin/
total 235080
lrwxrwxrwx    1 root     root             8 May 19 18:27 clang -> clang-14
lrwxrwxrwx    1 root     root             8 May 19 18:27 clang++ -> clang-14
-rwxr-xr-x    1 root     root      85718392 May 19 18:27 clang-14
lrwxrwxrwx    1 root     root             8 May 19 18:27 clang-cl -> clang-14
lrwxrwxrwx    1 root     root             8 May 19 18:27 clang-cpp -> clang-14
-rwxr-xr-x    1 root     root      42836776 May 19 18:27 clang-linker-wrapper
-rwxr-xr-x    1 root     root        500320 May 19 18:27 clang-nvlink-wrapper
-rwxr-xr-x    1 root     root       4417712 May 19 18:27 clang-offload-bundler
-rwxr-xr-x    1 root     root       3163312 May 19 18:27 clang-offload-wrapper
-rwxr-xr-x    1 root     root       1995504 May 19 18:23 clang-tblgen
lrwxrwxrwx    1 root     root             3 May 19 18:27 ld.lld -> lld
lrwxrwxrwx    1 root     root             3 May 19 18:27 ld64.lld -> lld
-rwxr-xr-x    1 root     root      56856976 May 19 18:27 lld
lrwxrwxrwx    1 root     root             3 May 19 18:27 lld-link -> lld
-rwxr-xr-x    1 root     root       8101840 May 19 18:26 llvm-ar
-rwxr-xr-x    1 root     root        227416 May 19 18:23 llvm-config
-rwxrwxr-x    1 root     root          1595 May 19 18:23 llvm-lit
-rwxr-xr-x    1 root     root       8646872 May 19 18:26 llvm-nm
-rwxr-xr-x    1 root     root       5027840 May 19 18:26 llvm-objcopy
-rwxr-xr-x    1 root     root      11420680 May 19 18:26 llvm-objdump
lrwxrwxrwx    1 root     root             7 May 19 18:26 llvm-ranlib -> llvm-ar
lrwxrwxrwx    1 root     root            12 May 19 18:26 llvm-readelf -> llvm-readobj
-rwxr-xr-x    1 root     root       7848232 May 19 18:26 llvm-readobj
lrwxrwxrwx    1 root     root            12 May 19 18:26 llvm-strip -> llvm-objcopy
-rwxr-xr-x    1 root     root       3916328 May 19 18:24 llvm-tblgen
lrwxrwxrwx    1 root     root             3 May 19 18:27 wasm-ld -> lld
```
@compnerd is there a ninja target for the llvm binutils substitute? I know you can query ninja, but I'm not very good at it.

Reply:
Otherwise it's `ninja install-llvm-nm install-llvm-ar install-objcopy ` etc. Let me see what's the minimal list necessary for stage 2 consumption. At least `llvm-ar` is.

Reply:
done. PTAL